electrum/gui/qt/email_notification_dialogs.py

```diff
@@ -248,31 +252,81 @@ def __init__(self, parent, email, error_msg=''):
         self.addSpacing(5)
         self.addWidget(self.error_label)
 
-    def resend_task(self):
-        t0 = datetime.datetime.now()
-        self.resend_strategy.resend()
-        t1 = datetime.datetime.now()
-        elapsed_time = (t1 - t0).total_seconds()
-        time.sleep(int(self.RESEND_COOL_DOWN_TIME - elapsed_time))
-
-    def on_success(self, *args, **kwargs):
-        self.resend_button.setEnabled(True)
-
     def on_error(self, errors):
         self.resend_strategy.on_error(errors[1])
         self.set_error(str(errors[1]))
-        self.resend_button.setEnabled(True)
 
     def resend_request(self):
         self.clear_error()
-        self.thread.add(task=self.resend_task,
-                        on_success=self.on_success,
-                        on_error=self.on_error)
-        self.resend_button.setEnabled(False)
+        self.cool_down_thread.run()
+        self.send_request_thread.add(
+            task=lambda: self.resend_strategy.resend(),
+            on_success=lambda *args, **kwargs: None,
+            on_error=self.on_error
+        )
 
     def pin(self):
         return self.input_edit.text()
 
+    def terminate_thread(self):
+        self.send_request_thread.terminate()
+
+
+class CoolDownThread:
+    RESEND_COOL_DOWN_TIME = 30
+
+    def __init__(self):
+        self.elapsed_time = 0
+        self.pin_layout = None
+        self.thread = None
+
+    def register_layout(self, layout: PinConfirmationLayout):
+        self.pin_layout = layout
+
+    def unregister_layout(self):
+        self.pin_layout = None
+
+    def _set_resend_button_status(self, text: str, enabled: bool):
+        if self.pin_layout:
+            self.pin_layout.resend_button.setText(text)
+            self.pin_layout.resend_button.setEnabled(enabled)
+
+    def get_left_time(self) -> int:
+        return int(self.RESEND_COOL_DOWN_TIME - self.elapsed_time)
+
+    def get_formatted_left_time(self) -> str:
+        return f'{self.get_left_time():>2}s'
+
+    def cool_down_task(self):
+        t0 = datetime.datetime.now()
+        while True:
+            self.elapsed_time = (datetime.datetime.now() - t0).total_seconds()
+            left_time = self.get_left_time()
+            self._set_resend_button_status(self.get_formatted_left_time(), False)
```

Comment:
Didnt test, but in this case dont we have an instant switch between labels `0` -> `resend`? Consider moving this between `if` and `sleep`.

Reply:
Logic was corrected.